src/platform/plugins/shared/discover/public/application/main/state_management/redux/internal_state.ts

```diff
@@ -266,43 +237,99 @@ export const internalStateSlice = createSlice({
     builder.addCase(loadDataViewList.fulfilled, (state, action) => {
       state.savedDataViews = action.payload;
     });
+
+    builder.addCase(initializeTabs.pending, (state) => {
+      state.tabs.areInitializing = true;
+    });
+
+    builder.addCase(initializeTabs.fulfilled, (state, action) => {
+      state.userId = action.payload.userId;
+      state.spaceId = action.payload.spaceId;
+      state.persistedDiscoverSession = action.payload.persistedDiscoverSession;
+    });
+
+    builder.addCase(saveDiscoverSession.fulfilled, (state, action) => {
+      if (action.payload.discoverSession) {
+        state.persistedDiscoverSession = action.payload.discoverSession;
+      }
+    });
+
+    builder.addMatcher(isAnyOf(initializeTabs.fulfilled, initializeTabs.rejected), (state) => {
+      state.tabs.areInitializing = false;
+    });
   },
 });
 
-const createMiddleware = ({
-  tabsStorageManager,
-  runtimeStateManager,
-}: {
-  tabsStorageManager: TabsStorageManager;
-  runtimeStateManager: RuntimeStateManager;
-}) => {
-  const listenerMiddleware = createListenerMiddleware();
+export const syncLocallyPersistedTabState = createAction<TabActionPayload>(
+  'internalState/syncLocallyPersistedTabState'
+);
+
+type InternalStateListenerEffect<
+  TActionCreator extends PayloadActionCreator<TPayload>,
+  TPayload = TActionCreator extends PayloadActionCreator<infer T> ? T : never
+> = ListenerEffect<
+  ReturnType<TActionCreator>,
+  DiscoverInternalState,
+  InternalStateDispatch,
+  InternalStateDependencies
+>;
```

Comment:
Our middleware listeners were untyped before and used `any` which made them difficult to update. Some explicit typing fixes that.